rules/clean-git-history/rule.md

```diff
@@ -0,0 +1,82 @@
+---
+type: rule
+title: Do you have a clean git history?
+uri: clean-git-history
+authors:
+  - title: Gordon Beeming
+    url: https://www.ssw.com.au/people/gordon-beeming/
+  - title: Daniel Mackay
+    url: https://ssw.com.au/people/daniel-mackay
+related:
+    - use-squash-and-merge-for-open-source-projects
+    - write-a-good-pull-request
+redirects: []
+created: 2024-02-22T00:00:00.000Z
+guid: af3e6656-97b6-424c-b608-be3dd004ba89
+---
+
+Maintaining a clean git history is important for readability and understanding the changes that have been made to a codebase. This is especially important when working in a team, as it allows you to see the changes that have been made to a file over time, and who made them. This can be useful for debugging, and for understanding why a particular change was made.
+
+Things that create a good git history include:
+
+<!--endintro-->
+
+- Granularity of commits
+- Descriptive commit messages
+- Easy to maintain (i.e. easily revert an entire feature)
+- Never lose history
+
+## Squashing Pull Requests
+
+`youtube: https://www.youtube.com/watch?v=0chZFIZLR_0`
+**Video: Git MERGE vs REBASE: Everything You Need to Know (4 min)**
+
+Squashing commits is the process of combining multiple commits into a single commit. This allows for the repository history to reflect changes to the application over time, rather than individual commits making up a feature which can be hard to follow as they give little context to the feature that the commit is being added for.
+
+## Good Pull Request titles
+
+When creating a pull request, it's important to have a clear and descriptive title. This is important when looking back at the git history, as it allows you to see at a glance what changes were made in a particular pull request.
```

Comment:
Should we add something about gitmoji?  I find helps to quickly scan a branch and see what sort of changes are going in.  

I think we have a rule about PBI titles.  Do we have one about PR titles?  Perhaps the PBI titles rule applies here too?

Reply:
We have a rule on PR titles, I've linked it in the related rules... I can move it inline here if that would be better

Reply:
As per our conversation, the writing good PRs rule in the related rules covers what's good for titles so we don't need to repeat it here